ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/Leaf.java

```diff
@@ -220,20 +269,55 @@ void optimize() {
+        availableNetworkActions = removeNetworkActionsIfMaxNumberReached(availableNetworkActions);
+        return availableNetworkActions;
+    }
+
+    /**
+     * Removes network actions far from most limiting element, using the user's parameters for activating/deactivating this
+     * feature, and setting the number of boundaries allowed between the netwrk action and the limiting element
+     *
+     * @param networkActionsToFilter: the set of network actions to reduce
+     * @return the reduced set of network actions
+     */
+    private Set<NetworkAction> removeNetworkActionsFarFromMostLimitingElement(Set<NetworkAction> networkActionsToFilter) {
         SearchTreeRaoParameters searchTreeRaoParameters = raoParameters.getExtension(SearchTreeRaoParameters.class);
         if (searchTreeRaoParameters.getSkipNetworkActionsFarFromMostLimitingElement()) {
             List<Optional<Country>> worstCnecLocation = getMostLimitingElementLocation();
-            return raoData.getAvailableNetworkActions()
-                    .stream()
-                    .filter(na -> !networkActions.contains(na)
-                            && isNetworkActionCloseToLocations(na, worstCnecLocation))
+            Set<NetworkAction> filteredNetworkActions = networkActionsToFilter.stream()
+                    .filter(na -> isNetworkActionCloseToLocations(na, worstCnecLocation))
                     .collect(Collectors.toSet());
+            if (networkActionsToFilter.size() > filteredNetworkActions.size()) {
+                LOGGER.debug("{} network actions have been filtered out because they are far from the most limiting element", networkActionsToFilter.size() - filteredNetworkActions.size());
+            }
+            return filteredNetworkActions;
         } else {
-            return raoData.getAvailableNetworkActions()
-                    .stream()
-                    .filter(na -> !networkActions.contains(na))
-                    .collect(Collectors.toSet());
+            return networkActionsToFilter;
+        }
+    }
+
+    /**
+     * Removes network actions for whom the maximum number of network actions has been reached
+     *
+     * @param networkActionsToFilter: the set of network actions to reduce
+     * @return the reduced set of network actions
+     */
+    Set<NetworkAction> removeNetworkActionsIfMaxNumberReached(Set<NetworkAction> networkActionsToFilter) {
+        Set<NetworkAction> filteredNetworkActions = new HashSet<>(networkActionsToFilter);
+        getMaxTopoPerTso().forEach((String tso, Integer maxTopo) -> {
+            long alreadyAppliedForTso = this.networkActions.stream().filter(networkAction -> networkAction.getOperator().equals(tso)).count();
+            if (alreadyAppliedForTso >= maxTopo) {
+                filteredNetworkActions.removeIf(networkAction -> networkAction.getOperator().equals(tso));
+            }
+        });
```

Comment:
shouldn't we also compare the maxRaPerTso with the alreadyAppliedTopo + alreadyAppliedRangeAction ?

`if(maxRaPerTso >= alreadyAppliedTopo + alreadyAppliedRangeAction)` we filter the network action from this TSO

Reply:
the already applied PSTs are deducted in method getMaxTopoPerTso()

Reply:
indeed, my bad